### DISCUSSION TITLE
Batch export for projects data and template findings

### DISCUSSION THREAD

**Discussion by noraj** [Syslifters/sysreptor#23] [General]
There is an export button on each project. But there is no export button to export all projects at once. Same for finding templates.

![image](https://github.com/Syslifters/sysreptor/assets/16578570/9d17d10c-f8ac-49f0-834e-4f0a0c70f5a7)
![image](https://github.com/Syslifters/sysreptor/assets/16578570/abc1f150-05f4-4826-b854-ad1f6f1a9de2)

**Comment by silent6trinity** [comment]
Bumping this, as being able to batch export & batch import the findings template(s) would be an enormous time saver.

**Reply by aronmolnar** [reply]
Batch export is still in our backlog, but we're not yet actively working on in.

However, batch import is already possible by combining multiple tar.gz files (this works with templates, designs, projects):

# Pack exported SysReptor items
You want to pack multiple projects/designs/templates to one tar.gz.

1. Put them into one directory
```bash
$ ls
'template-Cross-Site-Request-Forgery-(CSRF).tar.gz'               template-Session-management-weaknesses.tar.gz
 template-Disclosure-of-sensitive-data-in-URL-parameters.tar.gz  'template-Stored-Cross-Site-Scripting-(XSS).tar.gz'
 template-Incorrectly-configured-HTTP-security-headers.tar.gz     template-Untrusted-TLS-certificates.tar.gz
 template-Insecure-HTTP-cookies.tar.gz                            template-User-Enumeration.tar.gz
'template-SQL-Injection-(SQLi).tar.gz'                           'template-XML-External-Entity-Injection-(XXE).tar.gz'
```

2. Unpack, Repack
```bash
cat *.tar.gz | tar zxvf - -i && tar -czvf all.tar.gz *-assets *-images *.json 2>/dev/null || rm -rf *-assets *-images *.json 2>/dev/null
```

Your items will be packed to all.tar.gz


**Update:** Batch export is now possible via our cli tool (https://github.com/Syslifters/sysreptor/discussions/23#discussioncomment-7402403)

**Reply by noraj** [reply]
> Batch export is still in our backlog, but we're not yet actively working on in.

Should be quick and easy to implement.

The function could be a loop over the existing export function for each template (and having n `.tar.gz` of 1 `.json`) or a new function a bit different so we would have 1 `.tar.gz` with n `.json`.

A new root `/api/v1/findingtemplates/export/` (without UUID).

A new export button, calling the new export API endpoint.

![image](https://github.com/Syslifters/sysreptor/assets/16578570/3c0e9a47-6f07-4935-afe4-894fb97595f5)

**Comment by noraj** [comment]
## Temporary batch export workaround

I made a JavaScript script that will call the individual export API endpoint for each vulnerability template as a workaround waiting for the true batch export feature.

https://github.com/noraj/Pentest-collab-convert/blob/master/tools/sysreptor/export_all_vulns.js

cc @silent6trinity @aronmolnar

**Comment by aronmolnar** [comment]
Batch export of templates is now possible via reptor CLI.

```
pip3 install reptor
reptor conf
reptor template --export tar.gz
```

**Reply by noraj** [reply]
Nice feature but I fear I won't be able to use reptor much since I protected my sysreptor instance with a reverse proxy and client certificate authentication (eg. similar to what is explained in [this article](https://blog.raw.pm/en/Client-Side-Certificate-Authentication-in-Nginx-with-Elliptic-curve-cryptography/)). That's why having it in the WebUI would be nice as well.

**Reply by aronmolnar** [reply]
Good hint :)
Yes, it should also be in the webui.

And in reptor we could add the possibility to specify a client certificate.

**Comment by MWedl** [comment]
The web UI now supports bulk select and bulk export in list views

<img width="847" height="571" alt="image" src="https://github.com/user-attachments/assets/76a98965-8833-4670-8476-3fd127cd0491" />
